src/main/ApplicationUtils.cpp

```diff
@@ -131,6 +155,63 @@ runWithConfig(Config cfg, optional<CatchupConfiguration> cc)
     return 0;
 }
 
+int
+rebuildInMemoryLedger(Application& app)
+{
+    auto& ps = app.getPersistentState();
+    auto hasStr = ps.getState(PersistentState::kHistoryArchiveState);
+    auto lclHash = ps.getState(PersistentState::kLastClosedLedger);
+
+    HistoryArchiveState has;
+    has.fromString(hasStr);
+    // As the local HAS might have merges in progress, let
+    // `prepareForPublish` convert it into a "valid historical HAS".
+    has.prepareForPublish(app);
+
+    auto currentLedger =
+        LedgerHeaderUtils::loadByHash(app.getDatabase(), hexToBin256(lclHash));
+    if (!currentLedger)
+    {
+        throw std::runtime_error("Could not load ledger from database");
+    }
+
+    // Ensure all buckets are present on disk before application begins
+    for (auto const& b : has.currentBuckets)
+    {
+        auto checkBucket = [&](std::string bucketStr) {
+            auto bucketHash = hexToBin256(bucketStr);
+            auto bucket = app.getBucketManager().getBucketByHash(bucketHash);
+            if (!bucket || (isZero(bucket->getHash()) && !isZero(bucketHash)))
```

Comment:
Did you look at the implementation of `getBucketByHash`? It creates an empty bucket pointer if it can't find the bucket, so checking `!bucket` is not enough.

That being said, looking at the last known ledger loading code, we already check for missing buckets there, so I think I don't even need this check.

Reply:
> Did you look at the implementation of `getBucketByHash`? It creates an empty bucket pointer if it can't find the bucket, so checking !bucket is not enough.

yes I did, when it doesn't find the file, it does `return std::shared_ptr<Bucket>();` which is `nullptr` not an empty bucket.

other places in the code are only checking for `nullptr`

But yeah maybe we don't care anyways

Reply:
my bad, for some reason I was convinced that function uses `std::make_shared`. Sorry about that. But yeah, I removed this code now.